### PR TITLE
Support CA's priority expander config in `Shoot`s `Worker` spec

### LINKED ISSUE #10683: Specify ClusterAutoscaler's priority in shoot CR
**How to categorize this issue?**

/area auto-scaling
/kind enhancement

**What would you like to be added**:
Currently ClusterAutoscaler's priority configmap cluster-autoscaler-priority-expander is installed by user after cluster creation.
It should be part of shoot CR. 

**Why is this needed**:
We have more than one node groups to allow system components for cluster resilience.
But the cheapest one should take priority for TCO.
But we can not do that because configmap cluster-autoscaler-priority-expander is created by us after cluster ready.
I would like to make the configmap part of shoot yaml, so that gardener controller can create it in between control plane ready and system component deployment.

### PR DESCRIPTION
**Pull request by tobschli**
**How to categorize this PR?**

/area usability
/kind api-change

**What this PR does / why we need it**:
When creating a `Shoot`, it is currently not possible to directly pass the desired priority expander configuration ([ref](https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/expander/priority/readme.md?rgh-link-date=2024-11-19T15%3A24%3A32Z)) before the first node of a shoot is ready.

This PR adapts the `Shoot` spec in a way, that it is possible to specify the priority of`Worker` groups like the following:
```yaml
apiVersion: core.gardener.cloud/v1beta1
kind: Shoot
metadata:
  name: example-shoot
  namespace: garden-example
spec:
  provider:
    type: local
    workers:
    - name: local
      priority: 40 # priority of this worker group
      machine:
        type: local
    - name: local2
      machine:
        type: local
```

The field is optional.
When no priority is defined within the `Worker` groups, nothing happens and the CA gets deployed as usual.

Only when at least one priority is specified within the `Worker`s,  the `priority-expander` `ConfigMap` is generated and deployed to the `Shoot`.
This also entails, that the `priority` expander is enabled as the first expander in the CA configuration.
For every `Worker` group that has no priority defined, `0` is implied.

For the aforementioned `Shoot` definition, the generated `ConfigMap` will look like this:
```yaml
priorities: |
  "0":
  - shoot--local--local.shoot--local--local-local2
  "40":
  - shoot--local--local.shoot--local--local-local
```

**Which issue(s) this PR fixes**:
Part of #10683

**Special notes for your reviewer**:
As the passing of the priority information happens through the extension API, the extensions need to be revendored to support this feature.
**Release note**:

```feature user
It is now possible to specify the the priority of worker groups with the `Shoot.spec.provider.workers[].priority` field. When at least one priority is specified, the CA will respect this configuration before other expanders.
WARNING: When using this feature, Gardener will overwrite existing configurations that were made manual beforehand.
```

### CODE REVIEW

File: pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler.go [hunk 1]
```diff
@@ -449,8 +462,13 @@ func (c *clusterAutoscaler) computeCommand() []string {
 	}
 	gardencorev1beta1.SetDefaults_ClusterAutoscaler(c.config)
 
+	expanderMode := *c.config.Expander
+	if c.workersHavePriorityConfigured() {
```
Comment (plkokanov) [inline pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler.go]: You can assign this to a variable so that the loop is only executed once.

File: pkg/provider-local/controller/worker/machines.go [hunk 2]
```diff
@@ -145,13 +146,18 @@ func (w *workerDelegate) generateMachineConfig(ctx context.Context) error {
...
+		}
+
 		machineDeployments = append(machineDeployments, worker.MachineDeployment{
 			Name:                         deploymentName,
 			ClassName:                    className,
 			SecretName:                   className,
 			Minimum:                      pool.Minimum,
 			Maximum:                      pool.Maximum,
 			MaxSurge:                     pool.MaxSurge,
+			Priority:                     *pool.Priority,
```
Comment (plkokanov) [inline pkg/provider-local/controller/worker/machines.go]:
```suggestion
			Priority:                     ptr.Deref(pool.Priority, 0),
```

And you could remove the `if` check from line 149.

File: pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler.go [hunk 3]
```diff
@@ -449,8 +462,13 @@ func (c *clusterAutoscaler) computeCommand() []string {
...
+		expanderMode = "priority," + expanderMode
```
Comment (plkokanov) [inline pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler.go]: According to the cluster-autoscaler documentation

> an expander must not appear in the list more than once.

Since people are already using the `priority` expander: https://github.com/gardener/gardener/blob/fdce0541184cfb42c43b420f544b6ef5cb95f5d8/pkg/apis/core/v1beta1/types_shoot.go#L670
This could result in `expanderMode = priority,priority`

I'm not sure if specifying two expanders with the same name will lead to crashes, but if that's the case, we should perhaps reuse the `.clusterAutoscaler.expander` field and create the expander `ConfigMap` only if it is `priority`.

Alternatively, we could add validation to ensure that `expanderMode` is not `priority` when a worker pool has a value for its `.priority` field, which seems a bit misleading to me.

In any case, this should be properly documented.

File: pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler_test.go [hunk 4]
```diff
@@ -60,15 +60,32 @@ var _ = Describe("ClusterAutoscaler", func() {
...
+		machineDeployment2Name           = "pool2"
+		machineDeployment2Min      int32 = 3
+		machineDeployment2Max      int32 = 5
+		machineDeployment2Priority int32 = 40
+		machineDeployments               = []extensionsv1alpha1.MachineDeployment{
+			{Name: machineDeployment1Name, Minimum: machineDeployment1Min, Maximum: machineDeployment1Max, Priority: machineDeployment1Priority},
+			{Name: machineDeployment2Name, Minimum: machineDeployment2Min, Maximum: machineDeployment2Max, Priority: machineDeployment2Priority},
+		}
+
+		workerConfig = []gardencorev1beta1.Worker{
```
Comment (plkokanov) [inline pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler_test.go:90]: You could also add a test by passing a non-empty `workerConfig` that does not have priorities set.

**Review by plkokanov — CHANGES_REQUESTED** [review]
Nice PR!

What about also updating https://github.com/gardener/gardener/blob/fdce0541184cfb42c43b420f544b6ef5cb95f5d8/docs/usage/high-availability/shoot_high_availability_best_practices.md?plain=1#L66 to include the new way with which priorities can be specified?

File: pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler.go (hunk 3 above)
Comment (tobschli) [reply pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler.go]: Oh, yeah!

I am also not in favor of the validation approach, but on the other hand I also would not want to have the approach that priorities are only respected when the `expanderMode` is set to include the priority expander.

I also don't know a better approach unfortunately.
Maybe @rfranzke has one?

Comment (tobschli) [reply pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler.go]: @rfranzke proposed the idea of checking whether the user has already specified the `priority` expander in the expander option.
- If yes, we do not modify the CLI flag.
- In the case they didn't, we prepend the `priority` expander to the flag.

File: pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler.go [hunk 5]
```diff
@@ -617,11 +637,45 @@ func (c *clusterAutoscaler) computeShootResourcesData(serviceAccountName string)
...
+		Data: map[string]string{
+			"priorities": string(priorityConfig),
+		},
+	}
+
+	return configMap, nil
+}
+
+func ensureExpanderInExpanderConfig(expander string, expanderConfig gardencorev1beta1.ExpanderMode) gardencorev1beta1.ExpanderMode {
+	if strings.Contains(expander, string(expanderConfig)) {
```
Comment (plkokanov) [inline pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler.go]: This should be the other way around :)
```suggestion
	if strings.Contains(string(expanderConfig), expander) {
```

Could we also add a small test case to cover this function.

File: pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler.go [hunk 6]
```diff
@@ -617,11 +637,45 @@ func (c *clusterAutoscaler) computeShootResourcesData(serviceAccountName string)
...
+		}
+		objects = append(objects, configMap)
+	}
+	return registry.AddAllAndSerialize(objects...)
+}
+
+func (c *clusterAutoscaler) generatePriorityExpanderConfigMap() (*corev1.ConfigMap, error) {
+	priorities := map[int32][]string{}
+	for _, machineDeployment := range c.machineDeployments {
+		priorities[machineDeployment.Priority] = append(priorities[machineDeployment.Priority], fmt.Sprintf("%s.%s", c.namespace, machineDeployment.Name))
```
Comment (plkokanov) [inline pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler.go]:
```suggestion
		priorities[machineDeployment.Priority] = append(priorities[machineDeployment.Priority], fmt.Sprintf("%s\\.%s", c.namespace, machineDeployment.Name))
```

CA expects this to be a regular expression, so shouldn't the `.` be escaped?

**Review by plkokanov — CHANGES_REQUESTED** [review]
Sorry for the late re-review and thanks for addressing the comments so far 😊 

Could you add the new `spec.provider.workers[].priority` field as a comment to https://github.com/gardener/gardener/blob/1c6f30a0b8d0eb6cf67971893ec9dd3ecdc69b9b/example/90-shoot.yaml#L31

File: pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler.go (hunk 5 above)
Comment (tobschli) [reply pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler.go]: Sorry for that!

I see two options for creating this test case:
1. Export the function and explicitly test it
2. Introduce even more variables in the existing `Deploy` test.

Against the `1.` stands the argument that, other than testing, we do not _need_ to export the function.
But on the other hand `2.` is not a very beautiful solution as well, as the test is already quite complicated and this would make it even harder to understand IMO.

Am I missing a third option / what do you think?

Comment (plkokanov) [reply pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler.go]: I was thinking something more towards `2.`. You're right that the test has some complexity issues, but a possible change like this doesn't seem too complex, I think:
```diff
diff --git a/pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler_test.go b/pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler_test.go
index 52f0bb3fd6..19b4525e41 100644
--- a/pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler_test.go
+++ b/pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler_test.go
@@ -257,7 +257,7 @@ var _ = Describe("ClusterAutoscaler", func() {
                        },
                        Type: corev1.SecretTypeOpaque,
                }
-               deploymentFor = func(withConfig bool, withWorkerPriority bool) *appsv1.Deployment {
+               deploymentFor = func(withConfig, withWorkerPriority bool) *appsv1.Deployment {
                        var commandConfigFlags []string
 
                        expander := string(configExpander)
@@ -633,11 +633,14 @@ var _ = Describe("ClusterAutoscaler", func() {
 
        Describe("#Deploy", func() {
                Context("should successfully deploy all the resources", func() {
-                       test := func(withConfig bool, withWorkerConfig bool, runtimeVersionGreaterEquals126 bool) {
+                       test := func(withConfig bool, withWorkerConfig bool, runtimeVersionGreaterEquals126, withPriorityExpander bool) {
                                var config *gardencorev1beta1.ClusterAutoscaler
                                var shootWorkerConfig []gardencorev1beta1.Worker
                                if withConfig {
-                                       config = configFull
+                                       // Copy `configFull` so that the test doesn't overwrite it.
+                                       config = configFull.DeepCopy()
+                                       if withPriorityExpander {
+                                               config.Expander = ptr.To(gardencorev1beta1.ClusterAutoscalerExpanderPriority + "," + configExpander)
+                                       }
                                }
 
                                if withWorkerConfig {
@@ -709,11 +712,11 @@ var _ = Describe("ClusterAutoscaler", func() {
                                Expect(actualServiceMonitor).To(DeepEqual(serviceMonitor))
                        }
 
-                       It("w/o config", func() { test(false, false, false) })
-                       It("w/ config, kubernetes version < 1.26", func() { test(true, false, false) })
-                       It("w/ config, kubernetes version >= 1.26", func() { test(true, false, true) })
-                       It("w/ config, w/ workerConfig, kubernetes version >= 1.26", func() { test(true, true, true) })
-
+                       It("w/o config", func() { test(false, false, false, false) })
+                       It("w/ config, kubernetes version < 1.26", func() { test(true, false, false, false) })
+                       It("w/ config, kubernetes version >= 1.26", func() { test(true, false, true, false) })
+                       It("w/ config, w/ workerConfig, kubernetes version >= 1.26", func() { test(true, true, true, false) })
+                       It("w/ config, w/ workerConfig, kubernetes version >= 1.26, w/ 'priority' expander already configured", func() { test(true, true, true, true) })
                })
        })
```

Comment (tobschli) [reply pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler.go]: Alright, sounds good :)
Thank you for the insight!

File: pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler.go [hunk 7]
```diff
@@ -449,8 +464,13 @@ func (c *clusterAutoscaler) computeCommand() []string {
 	}
 	gardencorev1beta1.SetDefaults_ClusterAutoscaler(c.config)
 
+	expanderMode := *c.config.Expander
+	if workersHavePriorityConfigured {
+		expanderMode = ensureExpanderInExpanderConfig("priority", expanderMode)
```
Comment (plkokanov) [inline pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler.go]:
```suggestion
		expanderMode = ensureExpanderInExpanderConfig(gardencorev1beta1.ClusterAutoscalerExpanderPriority, expanderMode)
```

Let's use this also here.

File: pkg/component/extensions/crds/assets/crd-extensions.gardener.cloud_workers.yaml [hunk 8]
```diff
@@ -511,10 +516,16 @@ spec:
...
+                    priority:
+                      description: Priority (or weight) is the importance by which
+                        this machine deployment will be scaled by cluster autoscaling.
+                      format: int32
+                      type: integer
                   required:
                   - maximum
                   - minimum
                   - name
+                  - priority
```
Comment (plkokanov) [inline pkg/component/extensions/crds/assets/crd-extensions.gardener.cloud_workers.yaml]: Sorry that I missed this in the previous reviews. 
`priority` is set to required here which will cause an error during the reconciliation of the `Worker`, if the worker extensions are not deployed with support to manage the `priority` field. Basically, when the `MachineDeployment` status is added to the `Worker`'s status, it will not contain the `priority` field and the update will be rejected.

**Review by plkokanov — CHANGES_REQUESTED** [review]
What is the expectation if users specify `priority` for worker pools but  the corresponding worker extensions still do not support the required functionality? Is it ok for the priority expander `ConfigMap` to contain all 0s as priorities and also overwrite the user provided `ConfigMap`?

**Comment by tobschli** [comment]
> What is the expectation if users specify `priority` for worker pools but the corresponding worker extensions still do not support the required functionality? Is it ok for the priority expander `ConfigMap` to contain all 0s as priorities and also overwrite the user provided `ConfigMap`?

Oh yeah. The point in time where this is available in the Shoot spec, but not the extensions is probably problematic.
Let me talk to the team on how we can handle that.

**Comment by timuthy** [comment]
> > What is the expectation if users specify `priority` for worker pools but the corresponding worker extensions still do not support the required functionality? Is it ok for the priority expander `ConfigMap` to contain all 0s as priorities and also overwrite the user provided `ConfigMap`?
> 
> Oh yeah. The point in time where this is available in the Shoot spec, but not the extensions is probably problematic. Let me talk to the team on how we can handle that.

The `MachineDeployment` name is deterministic for all known provider extensions ([example](https://github.com/gardener/gardener-extension-provider-aws/blob/7450faae838450d6c1c78968bf19e33ead4f01a3/pkg/controller/worker/machines.go#L235C5-L235C21)).
I suggest that we build in a temporary fall-back that follows this scheme to find the configured priority from the `Shoot` resource, in case the priority is unset in the `Worker`. In case we can't find the priority, we might not put it into the config. WDYT?

**Comment by tobschli** [comment]
> The `MachineDeployment` name is deterministic for all known provider extensions ([example](https://github.com/gardener/gardener-extension-provider-aws/blob/7450faae838450d6c1c78968bf19e33ead4f01a3/pkg/controller/worker/machines.go#L235C5-L235C21)). I suggest that we build in a temporary fall-back that follows this scheme (`namespace` + "-" + `poolName`) to find the configured priority from the `Shoot` resource, in case the priority is unset in the `Worker`. In case we still can't find the priority, we might not put it into the config. WDYT?

There still is the problem that e.g. this feature gets released and users can specify priorities in their `Shoot` spec, but an the extension is not revendored yet.
I need to take a closer look at this scenario, but I think this will be problematic.

**Comment by tobschli** [comment]
> I suggest that we build in a temporary fall-back that follows this scheme (`namespace` + "-" + `poolName`) to find the configured priority from the `Shoot` resource, in case the priority is unset in the `Worker`. In case we still can't find the priority, we might not put it into the config. WDYT?

Actually we need to go the opposite way, I think.
We generate the `ConfigMap` when some priority is defined in the workers.

The ClusterAutoscaler deployment component gets both the `Worker` definition as well as the `MachineDeployments` that the extension built.

I made the Priority optional throughout all type definitions now, so we can allow the case that the extensions do not set the priority.

When we _know_ there should be priorities given to the `MachineDeployment`s, so when there are priorities defined in the workers and we do not see priorities, we need to extract the worker name from the `MachineDeployment`.

File: pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler.go [hunk 9]
```diff
@@ -617,11 +637,58 @@ func (c *clusterAutoscaler) computeShootResourcesData(serviceAccountName string)
...
+}
+
+func (c *clusterAutoscaler) generatePriorityExpanderConfigMap() (*corev1.ConfigMap, error) {
+	priorities := map[int32][]string{}
+	for _, machineDeployment := range c.machineDeployments {
+		// TODO(tobschli): Remove this once all well-known extensions have revendored to use the generic actuator that sets the priorities.
+		if machineDeployment.Priority == nil {
+			// In this case, the extension did not set the priorities that were configured in the worker.
+			// fall back to try to determine the pool name.
+			for _, pool := range c.workerConfig {
```
Comment (plkokanov) [inline pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler.go]: This double loop could be slow for clusters with large number of nodes. What about using a `map[string]int32` from `pool.Name` to `ptr.Deref(pool.Priority, 0)`, then getting the name of the pool from the `machineDeployment.Name` like this:
```
name := strings.TrimPrefix(machineDeployment.Name, c.namespace+"-")
zoneIndex := strings.LastIndex(name, "-z")
if zoneIndex != -1 {
    name = name[:zoneIndex]
}
``` 

and finally retrieving the priority from the map?

Note that since the `machineDeployment.Priority` is not explicitly defaulted to 0 anywhere (not in the provider-local extension and potentially not in other provider extensions later on) this branch will always be executed for worker pools that do not have `Priority` specified.

**Review by plkokanov — CHANGES_REQUESTED** [review]
Just one more optimisation comment

**Comment by tobschli** [comment]
In addition to revendoring the `generic actuator`, the `WorkerDelegates` of the provider extensions need to specifically pass the priority from the worker pools to the `MachineDeployments`.

I will create the respective follow-up PRs for it in the extensions.

**Review by plkokanov — COMMENTED** [review]
Thanks for addressing all my comments :) 
/lgtm

File: docs/usage/high-availability/shoot_high_availability_best_practices.md [hunk 10]
```diff
@@ -82,6 +82,22 @@ spec:
...
+spec:
+  provider:
+    workers:
+      - name: worker1
+        priority: 40 # priority of this worker group
+        machine:
+          type: local
+      - name: worker2
+        machine:
+          type: local
```
Comment (timuthy) [inline docs/usage/high-availability/shoot_high_availability_best_practices.md]:
```suggestion
    - name: worker1
      priority: 40 # priority of this worker group
      machine:
        type: local
    - name: worker2
      machine:
        type: local
```
Nit

File: docs/usage/high-availability/shoot_high_availability_best_practices.md [hunk 11]
```diff
@@ -82,6 +82,22 @@ spec:
       scaleDownUtilizationThreshold: 0.5
 ```
 
+In addition to that, it is also possible to configure the CA priority expander by adding priorities to the worker groups like this:
```
Comment (timuthy) [inline docs/usage/high-availability/shoot_high_availability_best_practices.md]: Could you explain the consequence of having a priority defined for only a few pools, i.e., what's the default prio for pools not having one?

Comment (timuthy) [inline docs/usage/high-availability/shoot_high_availability_best_practices.md]:
```suggestion
In addition to that, it is also possible to configure the [`cluster-autoscaler` priority expander](https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/expander/priority/readme.md) by adding priorities to the worker groups like this:
```

File: pkg/apis/core/types_shoot.go [hunk 12]
```diff
@@ -1213,6 +1213,8 @@ type Worker struct {
 	Sysctls map[string]string
 	// ClusterAutoscaler contains the cluster autoscaler configurations for the worker pool.
 	ClusterAutoscaler *ClusterAutoscalerOptions
+	// Priority (or weight) is the importance by which this worker group will be scaled by cluster autoscaling.
```
Comment (timuthy) [inline pkg/apis/core/types_shoot.go]:
```suggestion
	// Priority (or weight) is the importance by which this worker pool will be scaled by cluster autoscaling.
```

To remain consistent with other doc strings.

File: pkg/apis/core/v1beta1/types_shoot.go [hunk 13]
```diff
@@ -1580,6 +1580,9 @@ type Worker struct {
 	// ClusterAutoscaler contains the cluster autoscaler configurations for the worker pool.
 	// +optional
 	ClusterAutoscaler *ClusterAutoscalerOptions `json:"clusterAutoscaler,omitempty" protobuf:"bytes,21,opt,name=clusterAutoscaler"`
+	// Priority (or weight) is the importance by which this worker group will be scaled by cluster autoscaling.
+	// +optional
+	Priority *int32 `json:"priority,omitempty" protobuf:"varint,22,opt,name=priority"`
```
Comment (timuthy) [inline pkg/apis/core/v1beta1/types_shoot.go:1609]: The feature evolved a lot through discussions in the past and I couldn't follow any detail. However, I'm surprised that the `Priority` field is not part of the `ClusterAutoscalerOptions`. Was this a deliberate decision and if so, could you please link the reasoning, for my own education? 😉

File: pkg/apis/core/v1beta1/types_shoot.go [hunk 14]
```diff
@@ -1744,6 +1747,8 @@ var (
 	DefaultWorkerMaxUnavailable = intstr.FromInt32(0)
 	// DefaultWorkerSystemComponentsAllow is the default value for Worker AllowSystemComponents
 	DefaultWorkerSystemComponentsAllow = true
+	// DefaultWorkerPriority is the default value for Worker Priority
```
Comment (timuthy) [inline pkg/apis/core/v1beta1/types_shoot.go]:
```suggestion
	// DefaultWorkerPriority is the default value for Worker Priority.
```

File: pkg/apis/extensions/v1alpha1/types_worker.go [hunk 15]
```diff
@@ -252,4 +255,6 @@ type MachineDeployment struct {
 	Minimum int32 `json:"minimum"`
 	// Maximum is the maximum number for this machine deployment.
 	Maximum int32 `json:"maximum"`
+	// Priority (or weight) is the importance by which this machine deployment will be scaled by cluster autoscaling.
```
Comment (timuthy) [inline pkg/apis/extensions/v1alpha1/types_worker.go:258]:
```suggestion
	// Priority (or weight) is the importance by which this machine deployment will be scaled by cluster autoscaling.
	// +optional
```

File: pkg/apis/extensions/v1alpha1/types_worker.go [hunk 16]
```diff
@@ -164,6 +164,9 @@ type WorkerPool struct {
 	// ClusterAutoscaler contains the cluster autoscaler configurations for the worker pool.
 	// +optional
 	ClusterAutoscaler *ClusterAutoscalerOptions `json:"clusterAutoscaler,omitempty"`
+	// Priority (or weight) is the importance by which this worker pool will be scaled by cluster autoscaling.
+	// +optional
+	Priority *int32 `json:"priority"`
```
Comment (timuthy) [inline pkg/apis/extensions/v1alpha1/types_worker.go]:
```suggestion
	Priority *int32 `json:"priority,omitempty"`
```

File: pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler_test.go [hunk 17]
```diff
@@ -657,9 +713,12 @@ var _ = Describe("ClusterAutoscaler", func() {
...
 
-			It("w/o config", func() { test(false, false) })
-			It("w/ config, kubernetes version < 1.26", func() { test(true, false) })
-			It("w/ config, kubernetes version >= 1.26", func() { test(true, true) })
+			It("w/o config", func() { test(false, false, false, false) })
+			It("w/ config, kubernetes version < 1.26", func() { test(true, false, false, false) })
+			It("w/ config, kubernetes version >= 1.26", func() { test(true, false, true, false) })
+			It("w/ config, w/ workerConfig, kubernetes version >= 1.26", func() { test(true, true, true, false) })
+			It("w/ config, w/ workerConfig, kubernetes version >= 1.26, w/ 'priority' expander already configured", func() { test(true, true, true, true) })
+
```
Comment (timuthy) [inline pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler_test.go]:
```suggestion
```

File: pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler_test.go [hunk 18]
```diff
@@ -60,15 +60,41 @@ var _ = Describe("ClusterAutoscaler", func() {
...
+		machineDeployment3Name           = "pool3"
+		machineDeployment3Min      int32 = 3
+		machineDeployment3Max      int32 = 5
+		machineDeployments               = []extensionsv1alpha1.MachineDeployment{
+			{Name: machineDeployment1Name, Minimum: machineDeployment1Min, Maximum: machineDeployment1Max, Priority: machineDeployment1Priority},
+			{Name: machineDeployment2Name, Minimum: machineDeployment2Min, Maximum: machineDeployment2Max, Priority: machineDeployment2Priority},
+			{Name: machineDeployment3Name, Minimum: machineDeployment3Min, Maximum: machineDeployment3Max},
+		}
+
+		workerConfig = []gardencorev1beta1.Worker{
```
Comment (timuthy) [inline pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler_test.go:90]: Can you also test the fallback case, i.e., priority defined in shoot worker but not yet set in `MachineDeployment`?
In addition, please consider a test case in which the fallback value cannot be found.

File: pkg/apis/core/v1beta1/types_shoot.go [hunk 19]
```diff
@@ -1744,6 +1747,8 @@ var (
...
+	DefaultWorkerPriority = int32(0)
```
Comment (timuthy) [inline pkg/apis/core/v1beta1/types_shoot.go]: Just noticed: This one is unused. Leftover?

**Review by timuthy — CHANGES_REQUESTED** [review]
Thanks for the contribution 🙂

File: pkg/apis/core/v1beta1/types_shoot.go (hunk 13 above)
Comment (tobschli) [reply pkg/apis/core/v1beta1/types_shoot.go:1609]: The idea was to not introduce further coupling to the ClusterAutoscaler, as this is something that is not a detail of the ClusterAutoscaler, but used in a variety of autoscaling solutions.
Therefore I decided to not move it into the ClusterAutoscaler field.

Comment (timuthy) [reply pkg/apis/core/v1beta1/types_shoot.go:1609]: Fair enough, thanks 🙂

File: pkg/apis/extensions/v1alpha1/types_worker.go (hunk 15 above)
Comment (tobschli) [reply pkg/apis/extensions/v1alpha1/types_worker.go:258]: Is the newline for the `autoscaling` because of the lengthy string?
If yes, I would update this for the other ones as well.

Comment (timuthy) [reply pkg/apis/extensions/v1alpha1/types_worker.go:258]: Nope, only for `// +optional`.

Comment (tobschli) [reply pkg/apis/extensions/v1alpha1/types_worker.go:258]: Alright, thank you!

**Review by timuthy — COMMENTED** [review]
Nice!
/lgtm

**Review by timuthy — COMMENTED** [review]
/approve

**Comment by vlerenc** [comment]
Awesome, didn't notice, thank you @tobschli @timuthy. Which version of Gardener will contain the feature (and [your fix](https://github.com/gardener/gardener/issues/11272), if I see this right @tobschli)?

**Comment by tobschli** [comment]
> Awesome, didn't notice, thank you @tobschli @timuthy. Which version of Gardener will contain the feature (and [your fix](https://github.com/gardener/gardener/issues/11272), if I see this right @tobschli)?

v1.112 will contain the feature in the completely working state :)
I just confirmed it working during validation of the release.
